db/migration/schema_set.go

```diff
@@ -0,0 +1,219 @@
+// Copyright (C) 2020 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"database/sql"
+	"errors"
+)
+
+type stringSet map[string]bool
+
+type schemaSetBackend interface {
+	loadSchemaIDs() (stringSet, error)
+	persistSchemaID(schemaID string, tx *sql.Tx) error
+	deleteSchemaID(schemaID string) error
+}
+
+var errNilBackend = errors.New("schemaSetBackend is nil")
+
+type schemaSet struct {
+	schemas stringSet
+	backend schemaSetBackend
+}
+
+func newSchemaSet() *schemaSet {
+	return &schemaSet{
+		schemas: stringSet{},
+		backend: nil, // there is a global schemaSet instance, backend must be lazily initialized
+	}
+}
+
+func (s *schemaSet) init(backend schemaSetBackend) (err error) {
+	log.Notice("post-migration: initializing schema ID set")
+
+	if backend == nil {
+		return errNilBackend
+	}
+
+	s.backend = backend
+
+	schemas, err := backend.loadSchemaIDs()
+	if err != nil {
+		return err
+	}
+
+	s.schemas = schemas
+
+	return nil
+}
+
+func (s *schemaSet) markSchemaID(schemaID string, tx *sql.Tx) (err error) {
+	log.Info("post-migration: marking schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err = s.backend.persistSchemaID(schemaID, tx)
+	if err != nil {
+		return err
+	}
+
+	s.schemas[schemaID] = true
+	return nil
+}
+
+func (s *schemaSet) getSchemaIDs() []string {
+	schemas := make([]string, 0, len(s.schemas))
+	for schema := range s.schemas {
+		schemas = append(schemas, schema)
+	}
+	return schemas
+}
+
+func (s *schemaSet) removeSchemaID(schemaID string) error {
+	log.Info("post-migration: removing schema '%s'", schemaID)
+
+	if s.backend == nil {
+		return errNilBackend
+	}
+
+	err := s.backend.deleteSchemaID(schemaID)
+	if err != nil {
+		return err
+	}
+
+	delete(s.schemas, schemaID)
+	return nil
+}
+
+type postMigrationEventsBackend struct {
```

Comment:
where are the tests for this struct?

Reply:
I've introduced this struct in order to unit test program logic. There are only direct DB calls here and I'll add an integration test (unfortunately in `package cli`) to cover this.